### ISSUE TITLE
Repository analysis should not retry partial reads

### ISSUE THREAD

**Issue by DaveCTurner** [elastic/elasticsearch#101100]
Today `GetBlobChecksumAction` uses the same read path as for regular snapshot blob reads, which (for S3 and GCS repositories at least) will resume the download if it fails part-way through. Sometimes we will overwrite the blob while the read process is ongoing, and in that case the resumed download may see a different blob contents from the original. This is reported as a checksum mismatch, incorrectly indicating an atomicity failure in the repository. It would be better not to resume the download and instead to report that the analysis failed because the read was incomplete.

**Comment by ywangd** [comment]
Resuming read is _not_ a problem outside of repo analysis because we do _not_ overwrite blobs part way through?

**Comment by DaveCTurner** [comment]
There's several spots where an overwrite is possible, although they're certainly rare. It's essentially anywhere that the blob name is not made up randomly by the node that creates it. That includes:

- The root `RepositoryData` blob (but only if misconfigured to have multiple clusters writing to the same repo).
- The per-snapshot `SnapshotInfo` and `Metadata` blobs (if there's a master failover during finalization).
- The index-level `IndexMetadata` blobs, if using the pre-index-meta-gens repository format.
- The shard-level `BlobStoreIndexShardSnapshots` blobs, if using the pre-shard-gens repository format.
- The atomic register implementation, which I expect we'll use for `RepositoryData` consistency in a future version.

But also regardless of overwrites if we are seeing a high rate of partial reads then that's cause for concern in terms of S3-compatibility because we may not retry those reads forever and this will make snapshot restores unreliable. S3 is pretty robust in this regard (most of the time anyway) and today the repository analyzer will only pick up a problem in this area in the rare case of a blob overwrite and it will misreport it as a checksum failure. We should be more sensitive to partial reads, and definitely shouldn't report them as checksum failures.

**Comment by ywangd** [comment]
Thanks for the further clarification. IIUC, overwrite plus partial reads will lead to checksum failure. But it is not really a problem for actual s3 because it is robust in reading, i.e. partial reads do not really happen.

> if we are seeing a high rate of partial reads then that's cause for concern in terms of S3-compatibility
> We should be more sensitive to partial reads, and definitely shouldn't report them as checksum failures.

+1

**Comment by DaveCTurner** [comment]
>  But it is not really a problem for actual s3

Indeed, none of the things that the analyzer checks are a problem for actual S3, they're all ways in which some supposedly-S3-compatible storage implementation deviates from the behaviour of actual S3.
